File: leetcode/1971-find-if-path-exists-in-graph.py
# https://leetcode.com/problems/find-if-path-exists-in-graph/

from collections import defaultdict
from typing import List

class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        adj_list = defaultdict(list)
        for src, dest in edges:
            adj_list[src].append(dest)
            adj_list[dest].append(src)
        
        visited = set()
        def dfs(vertex):
            if vertex in visited: return False
            if vertex == destination: return True

            visited.add(vertex)
            for neighboor in adj_list[vertex]:
                if dfs(neighboor): return True
            
            return False

        return dfs(source)

File: leetcode/1971-find-if-path-exists-in-graph.js
// https://leetcode.com/problems/find-if-path-exists-in-graph/

/**
 * @param {number} n
 * @param {number[][]} edges
 * @param {number} source
 * @param {number} destination
 * @return {boolean}
 */
var validPath = function(n, edges, source, destination) {
  const adjList = {};
  for (let i = 0; i < n; i += 1) {
    adjList[i] = [];
  }

  for (let [src, dest] of edges) {
    adjList[src].push(dest);
    adjList[dest].push(src);
  }


  const visited = new Set();
  const dfs = (vertex) => {
    if (visited.has(vertex)) return false;
    if (vertex === destination) return true;

    visited.add(vertex);
    for (let neighboor of adjList[vertex]) {
      if (dfs(neighboor)) return true;
    }

    return false;
  }

  return dfs(source);
};
